### ISSUE TITLE
Question/Potential 3.1 Core Issue: Unable to cast object of type 'Microsoft.AspNetCore.Http.Endpoint' to type 'Microsoft.AspNetCore.Routing.RouteEndpoint'.

### ISSUE THREAD

**Issue by yahorsi** [Azure/azure-signalr#784]
Hi Guys,

We're trying to migrate our ASP.NET site to 3.1 from 2.2 and got several issues with Microsoft.Azure.SignalR. Currently, when we turn on Azure SignalR on dev machines we observe following exception.

```
System.InvalidCastException: Unable to cast object of type 'Microsoft.AspNetCore.Http.Endpoint' to type 'Microsoft.AspNetCore.Routing.RouteEndpoint'.
   at Microsoft.Azure.SignalR.NegotiateMatcherPolicy.CreateNegotiateEndpoint(Endpoint endpoint)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Azure.SignalR.NegotiateMatcherPolicy.ApplyAsync(HttpContext httpContext, CandidateSet candidates)
   at Microsoft.AspNetCore.Routing.Matching.DfaMatcher.SelectEndpointWithPoliciesAsync(HttpContext httpContext, IEndpointSelectorPolicy[] policies, CandidateSet candidateSet)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatch|8_1(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task matchTask)
   at EDW.CrewPage.Telemetry.RequestTelemetryMiddleware.Invoke(HttpContext context, ILogger`1 logger) in C:\Dev\1stquad\edw.crewpage\Source\EDW.CrewPage.Telemetry\RequestTimelineMiddleware.cs:line 71
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

Here is our code configurign services:
```CSharp
        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            services.AddDbContextPool<CrewPageContext>(opt =>
            {
                if (HostingEnvironment.IsDevelopment())
                {
                    opt.ConfigureWarnings(warnings => warnings.Throw(RelationalEventId.QueryClientEvaluationWarning));
                }
                opt.EnableSensitiveDataLogging();
            });
            services.AddTransient<IAzureSqlAuthTokenService, AzureSqlAuthTokenService>();

            services
                .AddMvc(options =>
                {
                    options.ModelBinderProviders.Insert(0, new EdwModelBinderProvider());
                    options.Filters.Add<MvcTimeFilter>();
                    options.Filters.Add<ModelValidationFilter>();
                    options.Filters.Add(new ResponseCacheAttribute
                    {
                        NoStore = true,
                        Location = ResponseCacheLocation.None
                    });
                })
                .AddControllersAsServices()
                /* 
                 * TODO: Move to the System.Text.Json serialization.
                 */
                .AddNewtonsoftJson(options => 
                {
                    options.SerializerSettings.DateTimeZoneHandling = DateTimeZoneHandling.Utc;
                    options.SerializerSettings.Converters.Add(new Newtonsoft.Json.Converters.StringEnumConverter());
                });

            services
                .AddSwaggerGen(c =>
                {
                    c.SwaggerDoc("v1", new OpenApiInfo { Title = "CrewPage API", Version = "v1" });

                    // See https://stackoverflow.com/questions/56234504/migrating-to-swashbuckle-aspnetcore-version-5
                    // First we define the security scheme
                    c.AddSecurityDefinition("Bearer", //Name the security scheme
                        new OpenApiSecurityScheme
                        {
                            Description = "JWT Authorization header using the Bearer scheme.",
                            Type = SecuritySchemeType.Http, //We set the scheme type to http since we're using bearer authentication
                            Scheme = "bearer" //The name of the HTTP Authorization scheme to be used in the Authorization header. In this case "bearer".
                        });

                    c.AddSecurityRequirement(new OpenApiSecurityRequirement
                    {
                        {
                            new OpenApiSecurityScheme{
                                Reference = new OpenApiReference{
                                    Id = "Bearer", //The name of the previously defined security scheme.
                                    Type = ReferenceType.SecurityScheme
                                }
                            },new List<string>()
                        }
                    });

                    var filePath = Path.Combine(PlatformServices.Default.Application.ApplicationBasePath, typeof(Startup).Namespace + ".xml");
                    c.IncludeXmlComments(filePath);

                    c.SchemaFilter<SwaggerExcludeFilter>();
                    c.OperationFilter<SwaggerExcludeFilter>();
                    c.SchemaFilter<SwaggerRequireValueTypePropertiesSchemaFilter>(/*camelCasePropertyNames:*/ true);

                    c.CustomSchemaIds(type => {
                        if (type.FullName.GetValueOrEmpty().StartsWith("EDW.CrewPage.Model.Database")){
                            throw new InvalidProgramException($"Type {type.FullName} is not allowed in API");
                        }

                        return type.Name;
                    });
                });


            services
                .AddAuthentication(o =>
                {
                    o.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
                })
                .AddEdwAzureAdAuth(o =>
                {
                    Configuration.GetSection("AzureAd").Bind(o);
                });

            services.AddAuthorizationPolicies(Configuration);

            if (!HostingEnvironment.IsDevelopment())
            {
                // Currently, Application Insights makes requests 2.5 times slower under debugging
                //
                // Note, AddApplicationInsightsTelemetry and UseApplicationInsights are intended to be mutually exclusive
                // So, if we add AI using services we should not use UseApplicationInsights for the WebHostBuilder
                // See https://github.com/Microsoft/ApplicationInsights-aspnetcore/wiki/Getting-Started
                services.AddApplicationInsightsTelemetry();
                services.AddApplicationInsightsTelemetryProcessor<SignalRFilterTelemetryProcessor>();
            }

            /* 
             * TODO: Move to the System.Text.Json serialization 
             */
            var signalRServiceBuilder = services.AddSignalR().AddNewtonsoftJsonProtocol(o =>
            {
                o.PayloadSerializerSettings.DateTimeZoneHandling = DateTimeZoneHandling.Utc;
                o.PayloadSerializerSettings.DateFormatHandling = DateFormatHandling.IsoDateFormat;
                o.PayloadSerializerSettings.Converters.Add(new Newtonsoft.Json.Converters.StringEnumConverter());
            });

            signalRServiceBuilder.AddAzureSignalR();

            services.Configure<AzureAdOptions>(options => Configuration.GetSection("AzureAd").Bind(options));
            services.AddAuthorization();
            services.AddMemoryCache();

            services.AddSingleton<ITelemetryInitializer, TelemetryInitializer>();

            services.AddMessageBus();
            services.AddSingleton<ISchedulerAccessor, SchedulerAccessor>();
            services.AddHostedService<MessageStoreService>();
            services.AddHostedService<SchedulerService>();
            services.AddHostedService<StatisticsService>();
            services.AddHostedService<PostponedValidationService>();
            services.AddHostedService<SignalRService>();
            services.AddHostedService<DatabaseEnumUpdaterService>();

            if (HostingEnvironment.IsStaging() || HostingEnvironment.IsDevelopment())
            {
                services.AddDataProtection().DisableAutomaticKeyGeneration();
            }

            return services.AddAutofacModules();
        }
```

And the one configures middlewares:
```CSharp
        public void Configure(
            IApplicationBuilder app,
            IWebHostEnvironment env, 
            IAzureSqlAuthTokenService azureSqlAuthTokenService)
        {
            app.UseRequestTelemetry(new RequestTelemetryMiddlewareOptions()
            {
                IsTimelineEnabled =
                    env.IsDevelopment() ||
                    env.IsStaging() ||
                    !string.IsNullOrEmpty(Configuration["ApplicationSettings:RequestTimeline"]),

                IsSecureToReturn =
                    (HttpContext context) =>
                    {
                        var result = context.User.HasClaim(x => 
                            x.Type == Constants.Claim.OriginalRole && 
                            Constants.Role.Administrators.Contains(x.Value));

                        return result;
                    }
            });

            app.UseStaticFiles(new StaticFileOptions
            {
                OnPrepareResponse = ctx =>
                {
                    const int durationInSeconds = 60 * 60 * 24;
                    ctx.Context.Response.Headers[HeaderNames.CacheControl] =
                        "public,max-age=" + durationInSeconds;
                }
            });
            app.UseRouting();

            if (env.IsDevelopment() || env.IsStaging())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
            }

            app.UseAuthentication();
            app.UseAuthorization();

            if (Configuration.GetValue("ApplicationSettings:EnableMaintenanceMode", true))
            {
                app.UseMaintenanceChecker();
            }

            if (env.IsProduction() || env.IsStaging())
            {
                //app.UseBrowserChecker();
            }

            DatabaseUpdater.Configuration = Configuration;
            DatabaseUpdater.Options.CommandTimeoutSeconds = 180;
            DatabaseUpdater.Options.StartUpdateAutomatically = !env.IsDevelopment();
            DatabaseUpdater.Options.OnConfigureConnection = (connection) =>
            {
                if (azureSqlAuthTokenService.IsEnabled)
                {
                    ConfigureAzureAuth(azureSqlAuthTokenService, (SqlConnection)connection);
                }
            };
            if (Configuration.GetValue("ApplicationSettings:EnableDatabaseUpdater", true))
            {
                app.UseDatabaseUpdater<DatabaseReadyService>();
            }

            app.UseSwagger(x=>x.SerializeAsV2 = true);
            app.UseSwaggerUI(c =>
            {
                c.SwaggerEndpoint("/swagger/v1/swagger.json", "CrewPage API V1");
            });

            app.UseMiddleware(typeof(ErrorHandlingMiddleware));

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                    name: "default",
                    pattern: "{controller}/{action=Index}/{id?}");

                endpoints.MapFallbackToController("index", "home");

                if (!IsScaleOutEnabled)
                    endpoints.MapHub<NotificationHub>("/hubs/notification");
            });
        }
```

The package version is
```
    <PackageReference Include="Microsoft.Azure.SignalR" Version="1.2.2" />
```

**Comment by yahorsi** [comment]
Just a few comments:

1. When I change code to:
1.1. Make a separate call to 
```CSharp
                app.UseAzureSignalR(routes =>
                {
                    routes.MapHub<NotificationHub>("/hubs/notification");
                });
```
1.2 Remove call endpoints.MapHub from app.UseEndpoints

The exception stops happening.
But, that is the opposite to what recommended in the various sources describing on how to migrate Azure SignalR Service to the 3.1

If I configure things as recommended:
2.1 Call AddAzureSignalR in configuring services
2.2 Remove explicit call to the UseAzureSignalR and leave just that:
```CSharp

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                    name: "default",
                    pattern: "{controller}/{action=Index}/{id?}");

                endpoints.MapFallbackToController("index", "home");

                if (!IsScaleOutEnabled)
                    endpoints.MapHub<NotificationHub>("/hubs/notification");
            });
``` 

I'm then getting the exception I have described in the initial post.

NOTE: There is UseRouting call

**Comment by JialinXin** [comment]
@yahorsi Thanks for reporting the issue. It's duplicated with #777 and we've already have a fix which will be soon released with version 1.2.3 in a day or two. Please try update to this version when it's available.

**Comment by JialinXin** [comment]
@yahorsi Please update to latest version 1.2.3 for the fix.
